[jit] New Iterable interface, refactor for loops (List, Tensor, Range)

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21666 [jit] Add for in enumerate to script
* **#21403 [jit] New Iterable interface, refactor for loops (List, Tensor, Range)**

wanchaol committed:
[jit] New Iterable interface, refactor for in list

wanchaol committed:
Update on "[jit] New Iterable interface, refactor for in list"

[jit] New Iterable interface, refactor for in list

gh-metadata: pytorch pytorch 21403 gh/wanchaol/15/head

wanchaol committed:
Update on "[jit] New Iterable interface, refactor for in list"

[jit] New Iterable interface, refactor for in list

gh-metadata: pytorch pytorch 21403 gh/wanchaol/15/head

Krovatkin commented on test/test_jit.py:
```diff
@@ -6261,6 +6261,17 @@ def fn():
         with self.assertRaisesRegex(torch.jit.Error, "must not be zero"):
             fn()
 
+    def test_script_for_in_list(self):
```
i think this is probably covered by `get_sum_list_fn` tests ( https://github.com/pytorch/pytorch/blob/master/test/test_jit.py#L3653 ) , but I might be mistaken.

Krovatkin reviewed (approved):
:shipit:

eellison reviewed (commented):
What would it take to land this PR without duplicating all of emitLoopCommon? We shouldn't have have a `emitNewLoopCommon` and a `emitLoopCommon`? 

Could you wait to land your stack until there is only one ?

wanchaol commented:
> What would it take to land this PR without duplicating all of emitLoopCommon? We shouldn't have have a `emitNewLoopCommon` and a `emitLoopCommon`?
> 
> Could you wait to land your stack until there is only one ?

@eellison Sure, `emitNewLoopCommon` is there to allow gradually transfer other for loops and easier reviews. We will only have one `emitLoopCommon` in the end, I will land the whole stack together once there's no duplicates.

zdevito commented on torch/csrc/jit/script/sugared_value.h:
```diff
@@ -94,6 +94,32 @@ struct TORCH_API SugaredValue
+  virtual std::vector<TypePtr> getItersTypeInfo(
+    const SourceRange& loc,
+    Function& m) {
+   throw ErrorReport(loc) << kind() << " does not have the iterables type information";
```
If these errors are user facing and need to describe what is happening in the same way python would if something could not be iterated.

zdevito commented on torch/csrc/jit/script/sugared_value.cpp:
```diff
@@ -236,6 +236,54 @@ std::shared_ptr<SugaredValue> SimpleValue::call(
+
+    // fill in the target element assignment value in the beginning of the FOR loop
+    {
+      TORCH_CHECK(iters_size == 1, "more than one iterable for in list");
```
This needs to be an error report, context is needed!

zdevito reviewed (commented):
Just a couple small things. What would it take to make sure we don't land extra loop emitting statements and do it in one PR? Elias is working on a lot of loop related stuff and it will slow his work down to have a lot of noise here.

wanchaol commented:
> Just a couple small things. What would it take to make sure we don't land extra loop emitting statements and do it in one PR? Elias is working on a lot of loop related stuff and it will slow his work down to have a lot of noise here.

@zdevito that requires migrating the for_range and while together, it requires extra work on hooking up Apply expressions, but it shouldn't be a problem. I will try to do them in this PR to avoid duplicating.

wanchaol committed:
add all for loops on "[jit] New Iterable interface, refactor for in list"

[jit] New Iterable interface, refactor existing for loops

gh-metadata: pytorch pytorch 21403 gh/wanchaol/15/head

wanchaol committed:
rebase with load/store and minor fix on "[jit] New Iterable interface, refactor for loops (List, Tensor, Range)"

[jit] New Iterable interface, refactor existing for loops

gh-metadata: pytorch pytorch 21403 gh/wanchaol/15/head

zdevito commented:
Let me know when the version working with the new Loop is ready to go. Maybe coordinate with Elias, since https://github.com/pytorch/pytorch/pull/21611 will also affect this logic.

wanchaol committed:
new iterable base on load/store on "[jit] New Iterable interface, refactor for loops (List, Tensor, Range)"

[jit] New Iterable interface, refactor existing for loops

gh-metadata: pytorch pytorch 21403 gh/wanchaol/15/head

wanchaol commented:
close in favor of https://github.com/pytorch/pytorch/pull/21801
